src/cage/RwaCageSettlement.sol

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// RwaInputConduit.sol -- In and out conduits for Dai
+//
+// Copyright (C) 2020-2021 Lev Livnev <lev@liv.nev.org.uk>
+// Copyright (C) 2021-2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import {DSTokenAbstract} from "dss-interfaces/dapp/DSTokenAbstract.sol";
+
+contract RwaCageSettlement {
+    uint256 internal constant WAD = 10**18;
+    uint256 internal immutable to18ConversionFactor;
+
+    DSTokenAbstract public immutable coin;
+    DSTokenAbstract public immutable gem;
+    uint256 public immutable price;
+
+    event Redeem(address indexed sender, uint256 gemWad, uint256 coinAmt);
+    event Deposit(address indexed sender, uint256 coinAmt);
+
+    constructor(
+        address _coin,
+        address _gem,
+        uint256 _price
+    ) public {
+        coin = DSTokenAbstract(_coin);
+        gem = DSTokenAbstract(_gem);
+        price = _price;
+
+        to18ConversionFactor = 10**uint256(18 - DSTokenAbstract(_coin).decimals());
+    }
+
+    function redeem(uint256 gemWad) external {
+        //   UNIT   =     (  WAD  *  WAD ) / (WAD *   (UNIT / WAD)   )
+        // cointAmt =     (gemWad * price) / (WAD * conversion_factor)
+        uint256 coinAmt = (gemWad * price) / (WAD * to18ConversionFactor);
+
+        require(gem.transferFrom(msg.sender, address(this), gemWad), "RwaCageSettlement/gem-transfer-failed");
+        require(coin.transfer(msg.sender, coinAmt), "RwaCageSettlement/coin-transfer-failed");
```

Comment:
A minor thing: maybe a more clear error message would be "insufficient amount of coins" or something?

Reply:
We would be assuming that the only reason `transferFrom()` fails is because of funding.
We can safely assume that for the gems, but not sure for coins.
Let's say this is used for USDC in the future and they blacklisted this contract. The tx would fail for other reason, so the message would be misleading.